detect when the UUID of the instance changes and reset all local caches when that occurs.

no2chem opened the pull request:
This fix should prevent a lockup issue when resetAll() is called.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/corfudb/corfudb/56)
<!-- Reviewable:end -->

no2chem commented:
@medhavidhawan 
<!-- Reviewable comment -K0YvL5amS2iJOzsOf0V:2097471484 -->

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/corfudb/corfudb/56)
<!-- Sent from Reviewable.io -->

medhavidhawan commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 46 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bRv03zR8VeEmEZmhe)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L46)):</sup>
would it be better to call this cachedViewUUID or something to that effect.
<!-- Reviewable comment -K0bRv03zR8VeEmEZmhf:1967989943 -->

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 142 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bOo7-Jaik02pdRU6l)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L142)):</sup>
can the view object be null ? If so this might throw NPE. 
<!-- Reviewable comment -K0bOo7-Jaik02pdRU6m:-202278504 -->

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 148 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bkBoWsQpioHqkUhmw)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L148)):</sup>
Would like a generic comment on when all this can happen. It would be helpful if this logic was separated out in a small method call as then it easier to write the comment as to when the view UUID  can change.
<!-- Reviewable comment -K0bkBoWsQpioHqkUhmx:-1339387699 -->

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 149 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bjcmITrIVorJOSzhe)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L149)):</sup>
Does the log.info need to be part of the synchronized block. Can we move it out by using some flag based on which we can log if we need to.
<!-- Reviewable comment -K0bjcmITrIVorJOSzhf:-699874789 -->

---

<sup>**[src/test/java/org/corfudb/runtime/view/StreamAddressSpaceIT.java, line 64 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bQxIn3R78SweL7GUB)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/test/java/org/corfudb/runtime/view/StreamAddressSpaceIT.java#L64)):</sup>
This statement is very dense. Needs comments. I cannot understand unless I understand underlying data structure.
<!-- Reviewable comment -K0bQxIn3R78SweL7GUC:-1149619297 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/corfudb/corfudb/56)
<!-- Sent from Reviewable.io -->

no2chem commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 46 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bRv03zR8VeEmEZmhe)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L46)):</sup>
Well, its actually the UUID of this view, so I think we should keep it as UUID.
<!-- Reviewable comment -K0cka3Aprtsjv1N_1vU:-1318590143 -->

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 142 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bOo7-Jaik02pdRU6l)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L142)):</sup>
View should not be null, because it is obtained from the CorfuDBRuntime which synchronously blocks until it gets a view.
<!-- Reviewable comment -K0ckx5pslCwIn1TBkeL:-1060153840 -->

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 148 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bkBoWsQpioHqkUhmw)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L148)):</sup>
Added in r3
<!-- Reviewable comment -K0cl481oRBO16fWcTYD:-1058992777 -->

---

<sup>**[src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java, line 149 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bjcmITrIVorJOSzhe)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/main/java/org/corfudb/runtime/view/LocalCorfuDBInstance.java#L149)):</sup>
Not sure why this is an issue. The log.info gets called only once, when the instance change is detected.
<!-- Reviewable comment -K0cl6rQeANBEJAt046M:1941606947 -->

---

<sup>**[src/test/java/org/corfudb/runtime/view/StreamAddressSpaceIT.java, line 64 \[r2\]](https://reviewable.io:443/reviews/corfudb/corfudb/56#-K0bQxIn3R78SweL7GUB)** ([raw file](https://github.com/corfudb/corfudb/blob/ba5848f1747392a43d304306ea840b929e1d0b0b/src/test/java/org/corfudb/runtime/view/StreamAddressSpaceIT.java#L64)):</sup>
fixed in r3.
<!-- Reviewable comment -K0clGGPucMVa_41jlMi:1981930627 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/corfudb/corfudb/56)
<!-- Sent from Reviewable.io -->

medhavidhawan commented:
Reviewed 3 of 4 files at r1, 2 of 2 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/corfudb/corfudb/56)
<!-- Sent from Reviewable.io -->
